design/workflow.go

```diff
@@ -0,0 +1,207 @@
+	})
+	Field(2, "created", String, "The workflow creation time", func() {
+		Format(FormatDateTime)
+		Example("2021-04-09T06:17:25Z")
+	})
+	Field(3, "name", String, "Name of the workflow", func() {
+		Example("TrainAndServe")
+	})
+	Field(4, "description", String, "Description for the workflow", func() {
+		Example("This workflow is just trains a model and serve it")
+	})
+	Field(5, "inputs", ArrayOf(WorkflowInput), "Inputs for the workflow")
+	Field(6, "outputs", ArrayOf(WorkflowOutput), "Outputs from the workflow")
+	Field(7, "steps", ArrayOf(WorkflowStep), "Steps to be executed by the workflow")
+
+	Required("name", "steps")
+})
+
+// WorkflowInput defines the input for a FuseML workflow
+var WorkflowInput = Type("WorkflowInput", func() {
+	Field(1, "name", String, "Name of the input", func() {
+		Example("mlflow-codeset")
+	})
+	Field(2, "description", String, "Description of the input", func() {
+		Example("An MLFlow project codeset")
+	})
+	Field(3, "type", String, "The type of the input (codeset, string, ...)", func() {
+		Example("codeset")
+	})
+	Field(4, "default", String, "Default value for the input", func() {
+		Example("mlflow-example")
+	})
+	Field(5, "labels", ArrayOf(String), "Labels associated with the input", func() {
+		Example([]string{"mlflow-project"})
+	})
+})
+
+// WorkflowOutput defines the output from a FuseML workflow
+var WorkflowOutput = Type("WorkflowOutput", func() {
+	Field(1, "name", String, "Name of the output", func() {
+		Example("prediction-url")
+	})
+	Field(2, "description", String, "Description of the output", func() {
+		Example("The URL where the exposed prediction service endpoint can be contacted to run predictions.")
+	})
+	Field(3, "type", String, "The data type of the output", func() {
+		Example("string")
+	})
+})
+
+// WorkflowStep defines a step for a FuseML workflow
+var WorkflowStep = Type("WorkflowStep", func() {
+	Field(1, "name", String, "The name of the step", func() {
+		Example("predictor")
+	})
+	Field(2, "image", String, "The image used to execute the step", func() {
+		Example("ghcr.io/fuseml/kfserving-predictor:1.0")
+	})
+	Field(3, "inputs", ArrayOf(WorkflowStepInput), "List of inputs for the step")
+	Field(4, "outputs", ArrayOf(WorkflowStepOutput), "List of output from the step")
+	Field(5, "env", ArrayOf(StepEnv), "List of environment variables available for the container running the step")
+})
+
+// WorkflowStepInput defines the input for a FuseML workflow step
+var WorkflowStepInput = Type("WorkflowStepInput", func() {
+	Field(1, "name", String, "Name of the input", func() {
+		Example("model-uri")
+	})
+	Field(2, "value", String, "Value of the input", func() {
+		Example("s3://mlflow-artifacts/3/c7ae3b0e6fd44b4b96f7066c66672551/artifacts/model")
+	})
+	Field(3, "codeset", StepInputCodeset, "Codeset associated with the input")
+})
+
+// StepInputCodeset defines the Codeset type of input for a FuseML workflow step
+var StepInputCodeset = Type("StepInputCodeset", func() {
```

Comment:
Hm, I'm thinking that currently codesets need to be identified by project in addition to the name. But if you are not using this structure yet, it might be fine just to keep it and fix it later...

Reply:
I agree.